### PR TITLE
Kernel state for Cell and Output

### PR DESCRIPTION
**Pull request by echarles**
This PR introduces a kernel state and let the Output and Cell components use that state via the Kernel Executor.

This allows more control on the ZMQ messages. The Notebook component will have to use that Kernel Executor also.

### CODE REVIEW

**Comment by echarles** [comment]
@sok82  I hope I have addressed the requirements of https://github.com/datalayer/jupyter-ui/issues/269 with a new ExecutionPhase state that goes informs when the Cell is completed. You can have a look at these videos and the Output and Cell examples of this PR. (it works even if no output is generated).

Screenshot from 2024-07-30 14-37-22

Screenshot from 2024-07-30 14-37-30

**Comment by echarles** [comment]
@MarcosVn Once this PR is merged, your PR https://github.com/datalayer/jupyter-ui/pull/280 will need to be revisited.

I have changed a bit how the cell is run in this PR, as well as the state. I have quickly looks the changes you propose, and overall looks good to me. To make your life easier, you can simply close 280 and open a brand new avoiding the merge pain (unless a rebase/merge works for you).

File: packages/react/src/components/output/Output.tsx [hunk 1]
```diff
@@ -85,29 +98,18 @@ export const Output = (props: IOutputProps) => {
...
           }
         });
       }
-      setAdapter(adapter);
-      outputStore.setAdapter(sourceId, adapter);
-      adapter.outputArea.model.changed.connect((outputModel, args) => {
-        setOutputs(outputModel.toJSON());
-      });
     }
   }, [id, kernel]);
```
Comment (fcollonval) [inline packages/react/src/components/output/Output.tsx:110]: Connected signal should be disconnected as part of the `useEffect` clean-up returned callback

File: packages/react/src/components/output/Output.tsx [hunk 2]
```diff
@@ -126,10 +128,10 @@ export const Output = (props: IOutputProps) => {
       };
     }
   }, [kernel]);
```
Comment (fcollonval) [inline packages/react/src/components/output/Output.tsx:130]: Same comment about signals

**Review by fcollonval — APPROVED** [review]
Thanks @echarles

**Comment by sok82** [comment]
> @sok82 I hope I have addressed the requirements of #269 with a new ExecutionPhase state that goes informs when the Cell is completed. You can have a look at these videos and the Output and Cell examples of this PR. (it works even if no output is generated).
> 
>  Screencast.from.07-30-2024.02.31.54.PM.webm 
>  Screencast.from.07-30-2024.02.34.33.PM.webm

Hey @echarles 

It's great to see progress on the issue, video looks promising.

Can you please provide few lines of code demonstrating how I can start execution of output in new model and react on changing of execution state (KernelPhase if I got you right)

Is tried to look at [this](https://github.com/datalayer/jupyter-ui/pull/282/files#diff-7067b9f57c7695da0272f2a6793eeae13aa1526a17f04c0bc81ba1b27ea194d7) but see no changes concerning reacting on execution state

**Comment by echarles** [comment]
> Is tried to look at [this](https://github.com/datalayer/jupyter-ui/pull/282/files#diff-7067b9f57c7695da0272f2a6793eeae13aa1526a17f04c0bc81ba1b27ea194d7) but see no changes concerning reacting on execution state

`{kernelsStore.getExecutionPhase(kernel.id)}` is what will return the phases, moving from RUNNING to COMPLETED.

**Comment by sok82** [comment]
> > Is tried to look at [this](https://github.com/datalayer/jupyter-ui/pull/282/files#diff-7067b9f57c7695da0272f2a6793eeae13aa1526a17f04c0bc81ba1b27ea194d7) but see no changes concerning reacting on execution state
> 
> `{kernelsStore.getExecutionPhase(kernel.id)}` is what will return the phases, moving from RUNNING to COMPLETED.

Great. Already found that.

And what about reacting on execution errors?

How can I check if execution finished with error?

**Comment by echarles** [comment]
> How can I check if execution finished with error?

Great call. For now, in case of error, I explicitly set COMPLETED. Let me change that to COMPLETED_WITH_ERROR

**Comment by echarles** [comment]
Like this?

[Screencast from 07-30-2024 05:46:04 PM.webm](https://github.com/user-attachments/assets/72cffb4b-20e0-4ed3-8c08-fa4d1a861fb5)

File: packages/react/src/components/output/Output.tsx (hunk 1 above)
Comment (echarles) [reply packages/react/src/components/output/Output.tsx:120]: Thx for @fcollonval for the review and help an this. Fixed with https://github.com/datalayer/jupyter-ui/pull/282/commits/771d78096376ba5f4d007ccc8e4bed81801236ff

File: packages/react/src/components/output/Output.tsx (hunk 2 above)
Comment (echarles) [reply packages/react/src/components/output/Output.tsx:140]: Thx for @fcollonval for the review and help an this. Fixed with https://github.com/datalayer/jupyter-ui/pull/282/commits/771d78096376ba5f4d007ccc8e4bed81801236ff

File: packages/react/src/components/output/Output.tsx [hunk 3]
```diff
@@ -54,60 +54,72 @@ export const Output = (props: IOutputProps) => {
...
+      }
+      adapter.outputArea.model.changed.connect(outputsCallback);
+      if (receipt) {
+        adapter.outputArea.model.changed.connect(receiptCallback)
+      }
+    }
+    return () => {
+      if (adapter) {
+        adapter.outputArea.model.changed.connect(outputsCallback);
+        adapter.outputArea.model.changed.connect(receiptCallback)
```
Comment (fcollonval) [inline packages/react/src/components/output/Output.tsx:117]:
```suggestion
        adapter.outputArea.model.changed.disconnect(outputsCallback);
        adapter.outputArea.model.changed.disconnect(receiptCallback)
```

**Comment by sok82** [comment]
Hey @echarles ,

I'm experimenting with version `0.17.0` having desired functionality and experiencing a problem

Kernel phase monitoring works fine when execution finishes successfully, but when something is wrong with a code exception is thrown from  kernel execution promise and I don't have an idea of how to catch it.

The way I trigger execution is just by increasing executionTrigger for output, so I don't control promise implicitly...

Exception example is 
```
Uncaught Error TypeError: bar() missing 1 required positional argument: 'height'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[12], line 1
----> 1 plt.bar(job_satisfaction_rate_train['job_satisfaction_rate'], color='green');

TypeError: bar() missing 1 required positional argument: 'height'
    at reject (/Users/sk/Develop/d20/d20-git-repo/d20-platform-wiki-editor-ui/node_modules/@lumino/coreutils/src/promise.ts:51:1)
    at _onReply (/Users/sk/Develop/src/jupyter/kernel/KernelExecutor.ts:262:1)
    at _handleReply (/Users/sk/Develop/d20/d20-git-repo/d20-platform-wiki-editor-ui/node_modules/@jupyterlab/services/src/kernel/future.ts:237:1)
    at handleMsg (/Users/sk/Develop/d20/d20-git-repo/d20-platform-wiki-editor-ui/node_modules/@jupyterlab/services/src/kernel/future.ts:219:1)
    at _handleMessage (/Users/sk/Develop/d20/d20-git-repo/d20-platform-wiki-editor-ui/node_modules/@jupyterlab/services/src/kernel/default.ts:1430:1)
    at <anonymous> (/Users/sk/Develop/d20/d20-git-repo/d20-platform-wiki-editor-ui/node_modules/@jupyterlab/services/src/kernel/default.ts:1595:1)
    --- Promise.then ---
    at KernelConnection._onWSMessage (/Users/sk/Develop/d20/d20-git-repo/d20-platform-wiki-editor-ui/node_modules/@jupyterlab/services/src/kernel/default.ts:1591:1)
```

Is there a chance to catch and process this exception?

What is a correct way of doing this?

**Comment by echarles** [comment]
@sok82 We could add more logic in this code block https://github.com/datalayer/jupyter-ui/blob/3ac5b261927f70a526f777b847ee84afc6ad4b3e/packages/react/src/jupyter/kernel/KernelExecutor.ts#L216-L224 so the state is notified in case of execution error. Worth opening a dedicated issue to track this feature.

**Comment by sok82** [comment]
Thanks for quick answer, @echarles !

We definitely need to handle this situation and not just explode with exception which is hard to handle

Opened https://github.com/datalayer/jupyter-ui/issues/289 to fix this.

In my case I just don't need this exception at all (because I have all the details in output which is enough), so for me it's better to suppress it (for example having some flag on output that will control suppression) in JS code and just pass exception description to output

**Comment by sok82** [comment]
Hey @echarles,
found another issue about controlling execution state - the one connected with output.

Actually what I need - is to get FINAL execution state AND output.

But in current implementation - these are two separate things, which could be obtained only separately.
1. Execution state - from Kernel execution phase
2. Output - from output state handler

And this is a problem for me in my flow of events which is
1. Trigger execution
2. Connect handler to output
3. Check that execution finished
4. Disconnect handler from output (this is important)

Previously I checked that execution is finished by receiving output change event and it worked, but I didn't knew in 100% cases state of execution - success or not

Now I am able to get the state, BUT when I receive new ExecutionPhase = completed I have to disconnect output handler - which is bad, because changing of execution phase can be triggered BEFORE output change is triggered. And in this case I do not catch output change event and lose new output value

I need some way to synchronize these two events, and for now I don't know how...

Can you please modify code of https://github.com/datalayer/jupyter-ui/blob/3ac5b261927f70a526f777b847ee84afc6ad4b3e/packages/react/src/jupyter/kernel/KernelExecutor.ts
in a way that solves this problem?
 
For me the optimal solution would be changing of execution phase ALWAYS before output change.
In this case I can check execution phase in output handler, knowing that it is final state and process output accordingly

**Comment by echarles** [comment]
Hi @sok82 would be great to open a separated issue for this. If you have an idea for a fix, an associated PR would be awesome also. Thx again for your reports.

**Pull request merged**